controllers/machine.go

```diff
@@ -82,45 +93,92 @@ func (mrc *machineReconcileContext) addFinalizer() error {
 	return nil
 }
 
-func (mrc *machineReconcileContext) ensureDependencies() error {
-	hardware, err := mrc.ensureHardware()
-	if err != nil {
-		return fmt.Errorf("ensuring hardware: %w", err)
-	}
+func (mrc *machineReconcileContext) isHardwareReady(hw *tinkv1.Hardware) bool {
+	return hw.Spec.Metadata.State == inUse && hw.Spec.Metadata.Instance.State == provisioned
+}
+
+type errRequeueRequested struct{}
+
+func (e *errRequeueRequested) Error() string {
+	return "requeue requested"
+}
+
+func (mrc *machineReconcileContext) ensureTemplateAndWorkflow(hw *tinkv1.Hardware) (*tinkv1.Workflow, error) {
+	wf, err := mrc.workflow()
+
+	switch {
+	case apierrors.IsNotFound(err):
+		if err := mrc.ensureTemplate(hw); err != nil {
+			return nil, fmt.Errorf("failed to ensure template: %w", err)
+		}
+
+		if err := mrc.createWorkflow(hw); err != nil {
+			return nil, fmt.Errorf("failed to create workflow: %w", err)
+		}
 
-	if err := mrc.ensureTemplate(hardware); err != nil {
-		return fmt.Errorf("ensuring template: %w", err)
+		return nil, &errRequeueRequested{}
+	case err != nil:
+		return nil, fmt.Errorf("failed to get workflow: %w", err)
+	default:
 	}
 
-	if err := mrc.ensureWorkflow(hardware); err != nil {
-		return fmt.Errorf("ensuring workflow: %w", err)
+	s := wf.GetCurrentActionState()
+	if s == tinkv1.WorkflowStateFailed || s == tinkv1.WorkflowStateTimeout {
+		mrc.log.Info("current action is in a failed or timed out state: %v", wf.GetCurrentAction())
 	}
 
-	return nil
+	return wf, nil
 }
 
-func (mrc *machineReconcileContext) markAsReady() error {
-	mrc.tinkerbellMachine.Status.Ready = true
+//
+func (mrc *machineReconcileContext) Reconcile() error {
+	defer func() {
```

Comment:
This defer for adding finalizer is going to be called each time the Machine reconcile loop runs.
Do want some checks to skip if the finalizer is already set?

Reply:
Would this be something we could come back around to? I feel like the scope of this initial concerns has grown quite a bit.

Reply:
Actually I did some digging into the until method. I think they handle that? Like even if the finalizer add is called multiple times it wont get duplicated :) Feel free to resolve the comment, thanks!